### DISCUSSION TITLE
Can I use X.509 certificates to identify a user with SSSD using "id_provider=files"

### DISCUSSION THREAD

**Discussion by buentead** [cockpit-project/cockpit#19421] [Q&A] [answered]
I'd like to avoid user/password on IoT gateways to login to Cockpit for security reasons. Hence, I tried to set up Cockpit for [Certificate authentication](https://cockpit-project.org/guide/latest/cert-authentication). I'd like to use the local Linux users, as the IoT gateways are not part of an MS-AD or simlar. The user name is in the CN of the certificate subject. Following a describtion of what I did so far.

The user certificate is signed by a CA, which I included in `/etc/sssd/pki/sssd_auth_ca_db.pem`.
The `/etc/sssd/sssd.conf` file content is:
```
[sssd]
enable_files_domain = true

[pam]
pam_cert_auth = True

[certmap/implicit_files/local_user]
matchrule = <ISSUER>.*CN=myIssuerCN.*
maprule = LDAPU1:(user={subject_dn_component.CN})
```
The /etc/cockpit/cockpit.conf contains:
```
[WebService]
ClientCertAuthentication = true
```

So when I call `https://<gateway-ip>:9090` the web-browser prompts me to select the certificate. Unfortunately, the cockpit-ws returns:
`cockpit-session: No matching user for certificate`

I couldn't find any helpful information how I could fix this issue. Hence my question here:
Is it possible to use X.509 certificates with SSSD id_provider=files (using local users) for Cockpit?

**Comment by buentead** [comment, marked as answer]
Yes, certificate authentication with just the local Linux users is possible. There is no central Identity Management System such as MS-AD required. The `/etc/sssd/sssd.conf` file, however, need to be different according to the design page [Certificate mapping and matching rules for all providers](https://sssd.io/design-pages/certmaps_for_LDAP_AD_file.html). The following example works for me as the required user 'admin' is in the CN of the subject:
```
[SSSD]
enable_files_domain = true

[domain/implicit_files]
id_provider=files

[certmap/implicit_files/admin]
matchrule = <SUBJECT>^.*CN=admin.*$
```

Important is the rule name within the `certmap` section, which is the Linux user, 'admin' in the example above. The `matchrule` is used  to assign a certificate content to this user. Whereas the validity of the signed certificate is already checked against the `/etc/sssd/pki/sssd_auth_ca_db.pem` CA chain.

**Reply by saner20** [reply]
For some reason this is still not working for me. Cockpit just takes me back to the login page with a 401 error. My CN match is last.first.middle.123456. my `matchrule = <SUBJECT>^.*CN=LAST.FIRST.MIDDLE.123456.*$` and my sssd_auth_ca_db.pem has my root and intermediate certificates in pem format.

**Reply by buentead** [reply]
Did you create the user in Linux that is assigned in the 'certmap' statement? In my example it is [certmap/implicit_files/admin] and the Linux user 'admin' must exist.
The regex you have is case sensitive. Hence, make sure upper- / lowercase match.
_Remark:_ If you really want to check for the dots in the common name, you need to escape them like "^.*CN=LAST\.FIRST\.MIDDLE.*$". Otherwise it will accept any character where the dot is.

**Reply by saner20** [reply]
yes, the user does exist in my passwd file. Below is my current configuration. Is the <Subject> standard or should that match my certificate?

`[sssd]
enable_files_domain = true
debug_level = 10`
`[domain/localDomain]
id_provider = files`
`[certmap/localDomain/admin]
matchrule = <Subject>^.CN=LAST.FIRST.MIDDLE.123456.$`

**Reply by saner20** [reply]
the error I see in messages log
` cockpit-session: Failed to map certificate to user: [org.freedesktop.DBus.Error.ServiceUnknown] The name org.freedesktop.sssd.infopipe was not provided by any .service files`

**Reply by martinpitt** [reply]
@saner20 You are most probably missing the `sssd-dbus` package -- at least that's the name in Fedora/RHEL/Debian.

**Reply by saner20** [reply]
@martinpitt yes, after running rpm -qa | grep ssd that package is not on my system. Is that all I need to install to successfully map certificates?

**Reply by martinpitt** [reply]
@saner20 Package wise that should be all. Config-wise I don't know, but should be.

**Reply by saner20** [reply]
Adding sssd-dbus package was an issue and it has helped with logging, but now I am getting an error from the p11_child.log and the sssd_ifp.log. I am not positive if it's because my pem isn't formatted properly or because my certmap configuration in the sssd.conf is incorrect. 
`(0x0040): X509_LOOKUP_load_file [/etc/sssd/pki/sssd_auth_ca_db.pem] failed [33558541][error:0200100D:system library:fopen:Permission denied].`
` [ifp] [ifp_users_find_by_valid_cert_step] (0x0040): /usr/libexec/sssd/p11_child failed with status [1]. Check p11_child logs for more information.
   *  (2024-08-14 15:43:40): [ifp] [sbus_issue_request_done] (0x0040): org.freedesktop.sssd.infopipe.Users.FindByValidCertificate: Error [1432158298]: Invalid certificate provided`

**Reply by saner20** [reply]
My issues was selinux related. SELinux was preventing access to sssd_auth_ca_db.pem. 
`SELinux is preventing /usr/libexec/sssd/p11_child from read access on the file /etc/sssd/pki/sssd_auth_ca_db.pem`
